providers/dns/plesk/internal/client.go

```diff
@@ -0,0 +1,136 @@
+package internal
+
+import (
+	"bytes"
+	"encoding/xml"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+)
+
+// Client the Plesk API client.
+type Client struct {
+	HTTPClient *http.Client
+	baseURL    *url.URL
+	login      string
+	password   string
+}
+
+// NewClient created a new Client.
+func NewClient(baseURL *url.URL, login string, password string) *Client {
+	return &Client{
+		HTTPClient: &http.Client{Timeout: 10 * time.Second},
+		baseURL:    baseURL,
+		login:      login,
+		password:   password,
+	}
+}
+
+// GetSite gets a site.
+// https://docs.plesk.com/en-US/obsidian/api-rpc/about-xml-api/reference/managing-sites-domains/getting-information-about-sites.66583/
+func (c Client) GetSite(domain string) (int, error) {
+	payload := RequestPacketType{Site: &SiteTypeRequest{Get: SiteGetRequest{Filter: &SiteFilterType{
+		Name: domain,
+	}}}}
+
+	response, err := c.do(payload)
+	if err != nil {
+		return 0, err
+	}
+
+	if response.System != nil {
+		return 0, response.System
+	}
+
+	if response.Site.Get.Result.Status != StatusOK {
```

Comment:
I feel like this is a tad dangerous. Since `response.Site` is a pointer (and could be nil), accessing `response.Site.Get` may panic.

We can assume the Plesk server will behave correctly, but I'd feel better when we add explicit nil checks here.

Background: Back in 2009 or 2010 (IIRC) I was integrating some Java software for machine provisioning with Plesk, and it didn't always produce the documented response (often missing fields altogether and even returning bogus data). One such hickup managed to kill the hoster's provisioning queue for a few days, and identifying the job responsible took forever.

I'm sure (I hope!) the situation today is way better, but Plesk is still implemented in PHP, so I'm hesitant to place trust in that software...

Reply:
I was not comfortable with that, so I agree with you